### DISCUSSION TITLE
How to negate a condition in "route_rule.data"?

### DISCUSSION THREAD

**Discussion by curiosport** [bfenetworks/bfe#1045] [Q&A (问题和回答)] [answered]
I have this condition:
```"Cond": "req_path_prefix_in(`/register|/recover|/swap`, false)"```

And I can't deny it, I have used several combinations using the ! symbol and none of them work:
```
"Cond": "! req_path_prefix_in(`/register|/recover|/swap`, false)"
"Cond": "!req_path_prefix_in(`/register|/recover|/swap`, false)"
"Cond": "!(req_path_prefix_in(`/register|/recover|/swap`, false)")
"Cond": ! "req_path_prefix_in(`/register|/recover|/swap`, false)"
"Cond": !"req_path_prefix_in(`/register|/recover|/swap`, false)"
```

This is the file:
```
{
	"Version": "1",
	"ProductRule": {
		"example_product": [
			{
				"Cond": "req_cookie_key_in(`sample-session`)",
				"ClusterName": "cluster_sample_page"
			},
			{
				"Cond": "!req_path_prefix_in(`/register|/recover|/swap`, false)",
				"ClusterName": "cluster_sample_page"
			},
			{
				"Cond": "default_t()",
				"ClusterName": "cluster_other_page"
			}
		]
	}
}
```

By the way, in the official documentation you can find it in this format:
```"Cond: "req_cookie_key_in("sample-session")"```

However that doesn't work and throws an error during JSON parsing.

These forms work:
```"Cond: "req_cookie_key_in(\"sample-session\")"```
```"Cond: "req_cookie_key_in(`sample-session`)"```

My suggestion is that they update the documentation or at least point it out.

**Comment by daimg** [comment, marked as answer]
Condition does support Logical NOT operator "!".
The Expression:
```
"Cond": "!req_path_prefix_in(`/register|/recover|/swap`, false)",
```

would reject any path which is prefixed with /register, /recover  , or /swap, for example, /register /registerabc /register/abc, etc. What's path you used in the test?

And in your conf file, primitive "default_t()" is used, so any request which doesn't match following conditions would fall back to cluster "cluster_other_page"

```
"Cond": "req_cookie_key_in(`sample-session`)",
"Cond": "!req_path_prefix_in(`/register|/recover|/swap`, false)",
```



> By the way, in the official documentation you can find it in this format:
> "Cond: "req_cookie_key_in("sample-session")"
> 
> However that doesn't work and throws an error during JSON parsing.
> 
> These forms work:
> "Cond: "req_cookie_key_in(\"sample-session\")"
> "Cond: "req_cookie_key_in(`sample-session`)"
> 
> My suggestion is that they update the documentation or at least point it out.

Yes, the embedded quotes should be escaped for json parsing. We may update the doc later. Thanks for the suggestion.

**Reply by curiosport** [reply]
Thanks for your reply, it made me reconsider my code and it turns out this was the solution for my case:
```"Cond": "!req_path_prefix_in(`/|/register|/recover|/swap`, false)",```

I was missing also denying the root that's why I was so confused, sorry.

**Reply by curiosport** [reply]
I'm doing a lot of things and didn't realize it at the time but I just noticed that my condition really doesn't help, my intention is to deny the following paths:
```
https://127.0.0.1:8555/
https://127.0.0.1:8555/register
https://127.0.0.1:8555/recover
https://127.0.0.1:8555/swap
```

Apparently my condition denies all paths due to ```(`/|``` that I wrote it with the intention of denying the root path.

Can you please tell me what to do?

**Reply by daimg** [reply]
To match path exactly, you may use req_path_in(), rather than req_path_prefix_in().
```
"Cond": "!req_path_in(`/|/register|/recover|/swap`, false)",
```

Refer to docs: https://www.bfe-networks.net/en_us/condition/request/uri/

**Reply by curiosport** [reply]
Oh! I didn't mention it, but these should also be denied:
```
https://127.0.0.1:8555/swap/blabla
https://127.0.0.1:8555/swap/etc
```
Using two conditions is the only way right?
```
"Cond": "!req_path_prefix_in(`/swap`, false)",
"Cond": "!req_path_in(`/|/register|/recover`, false)",
```

**Reply by daimg** [reply]
You don't need two separated  conds, some operators can be used to define the condition expression you expect. 
See more details in  https://www.bfe-networks.net/en_us/condition/condition_grammar/.   (Advanced Condition Expression)

**Reply by curiosport** [reply]
bfe supports multiline conditions?
something like this:
```
"Cond": "req_cookie_key_in(`sample-session`)
    | || !req_path_prefix_in(`/swap`, false)
    | || !req_path_in(`/|/register|/recover`, false)",
```
I tried the above because that's how I saw it in the documentation but it didn't work, either I'm using the wrong `|` symbol or there really is no support.
